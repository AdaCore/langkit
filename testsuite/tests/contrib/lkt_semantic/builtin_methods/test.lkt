class Name : Node implements TokenNode {

    fun parent_node (): Node = node.parent

    fun parent_env (): LexicalEnv = node.parent.node_env()
    fun parent_env (): LexicalEnv = node.parent_node().node_env()

    fun parent_env_node (): Node = node.parent.node_env().get(node.symbol)?(0)
    fun parent_env_node (): Node = node.parent_env().get(node.symbol)?(0)

    fun analysis_unit_node (): Node = node.unit().root

    fun first_node (): Node = node.node_env().get_first(node.symbol)

    fun env_node (): Node = node.node_env().env_node()

    fun children_env (): LexicalEnv = node.children_env()

    fun to_symbol (s : String): Symbol = s.to_symbol

    fun parent_nodes (): Array[Node] = node.parents()
    fun parent_nodes (): Array[Node] = node.parents(with_self=false)

    fun node_iterator (): Iterator[Node] = node.parents().to_iterator

}
