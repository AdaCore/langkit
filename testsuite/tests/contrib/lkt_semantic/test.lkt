@root_node class A {
    fun identity(): A = self

    fun test_int(): Int = 12

    fun test_int_invalid(): A = 12

    fun test_int_add(): Int = 12 + 15

    fun test_int_add_invalid_1(): Int = "12" + 15

    fun test_int_add_invalid_2(): Int = 15 + self

    fun test_int_add_invalid_3(): A = 15 + self

    fun test_regexp_1(): Regexp = p"[a-z]"

    fun test_regexp_2(): Regexp = {
        val rg = p"[A-Z]";
        rg
    }

    fun test_if_1(): Int = if true then 1 else 2

    fun test_if_invalid_1(): Int = if true then 1 else "2"
}

generic [C] struct Pair {
    a : C
    b : C
}

generic [C, D] struct Tuple2 {
    a : C
    b : D

    fun test(): C
}

class B : A {
    fun test_generic_constructor(): Pair[Int] = Pair[Int](12, 15)
    fun test_generic_component_access(): Int = Pair[Int](12,  15).a
    fun test_invalid_generic_type(): Pair[Int] = Pair[Int](12, 15).a
    fun test_generic_component_access_2(): Bool = Tuple2[Int, Bool](12, true).b
    fun test_invalid_generic_type_2(): Tuple2[Int, Bool] = true
    fun test_fun_call(arg : A): Int = arg.test_int_add()
    fun test_struct_fun(arg: Tuple2[Int, Bool]): Int = arg.test()
    fun test_array_indexing(arg: Array[Int]): Int = arg (12)
}

val a : Int = 12

struct Foo {
    # Test that you cannot access toplevel declarations from a nested scope.
    fun test_identifier_leak_invalid(): Int = self.a
}

enum class C (D, E) {
    fun wot(): D
}

# Test function types
fun add (a : Int, b : Int): Int = a + b
fun get_fn(): (Int, Int) -> Int = add
fun get_method(b_inst: B): (B) -> Int = b_inst.test_generic_component_access
fun get_method_invalid(b_inst: B): (B) -> Int = b_inst.test_fun_call

# Test Array/ASTList primitives

fun test_astlist_len(self: ASTList[String]): Int = self.length()
fun test_array_len(self: Array[String]): Int = self.length()
fun test_astlist_indexing(self: ASTList[String]): String = self (12)

# Test that errors flowing from a subtree to its children don't raise an
# exception

# Here, self doesn't exist. Since we're going to try to resolve the type of
# `12` first, it's going to need to resolve `self, which doesn't exist. This
# shouldn't crash.
fun test_astlist_indexing(a: ASTList[String]): String = self (12)

# Test traits, part 1: test that trait methods are visible in types
# implementing them.

trait WithLength {
    @builtin fun length(): Int
}

generic [T] trait WithIndex {
    @builtin fun __call__(index : Int): T
}

generic [T] class Vec implements WithLength, WithIndex[T] {
}

val vec : Vec[String] = null
val vec_length : Int = vec.length()
val vec_item : String = vec (1)

# Test resolution of named parameters

fun test_named_parameters(a: Int): Int = a
val ss : Int = test_named_parameters(a=12)

# Test resolution of enum class subclasses (should be accessible via dot
# notation on the root enum class type).

enum class EnClass (SubclassA, SubclassB, SubclassC) {}

val en_class_inst : EnClass.SubclassB = EnClass.SubclassB()
# Test that those classes don't leak
val invalid_en_class : SubclassB = SubclassB()

# Test errors for mismatched actuals & formals in function calls

fun test(a: Int, b: Int): Int = a + b

val t : Int = test(12, 15) # Correct call
val u : Int = test(12) # Missing params
val v : Int = test(b=12) # Missing params in call with kwargs
val x : Int = test(12, 15, 17) # Too many params
val y : Int = test(a=1, b=12, c=15) # Too many params kwargs
val z : Int = test(a=1, b=12, c=foo) # Invalid kwarg with wrong value: check that there is an error message for the expr

# Regression test for semantic failure over dotted names

val a : Int = 1
val b : Int = a.plus(2)

 # Test that LKNode.bool_type resolves correctly
fun test_if_2(a: Int, b: Int): Int = if a = b then a else b

#
# Test lambda expressions and parameters
#

fun lambda_test_helper(int: Int, fn: (Int, Int) -> Int): Int

# Test 1: test that type of parameters is correctly infered
fun lambda_test_1(): Int = lambda_test_helper(12, (x, y) => x + y * 2)

# Test 2: test that either all or no lambda params have type annotations
fun lambda_test_incorrect_1(): Int = lambda_test_helper(12, (x : Int, y) => x + y * 2)

# Test 3: Errors on mismatching lambda type
fun lambda_test_incorrect_2(): Int = lambda_test_helper(12, (x : String, y : String): Int => 12)

#
# Test availability of base type's methods
#

class Foo {
    fun bar(): Int
}

class Baz : Foo {}

fun test_base_type_methods(arg: Baz): Int = arg.bar()

#
# Generic functions instantiation tests
#

# Support generics
generic [T] fun convert_int(var: Int): T
generic [T] fun identity(var: T): T
generic [T] fun singleton(var: T): Array[T]
generic [T] fun apply (fn: (T) -> T, v: T): T

# Explicit instantiations
fun test_explicit_int_to(): String = convert_int[String](12)
fun test_explicit_identity(): Int = identity[Int](12)
fun test_explicit_singleton(): Array[Int] = singleton[Int](12)

# Explicit generic instantiation with function argument
fun double (arg: Int): Int = arg * 2
fun test_apply(): Int = apply[Int](double, 12)

# Test paren expr
fun test_paren_expr(): Int = (if true then 12 else null)

# Test null exclusion expr
fun test_null_excl_expr(): Int = (if true then 12 else null)!

# Implicit instantiations

# Check that we can instantiate a function depending on its target type,
# without any dependent parameters.
fun test_int_to(): String = convert_int(12)

# Check that we can instantiate a function depending on its target type, and
# properly resolve parameters with types that depend on the generic formal
# types.
fun test_identity(): Int = identity(12)

# Check that we can instantiate a function with a more complex return type that
# is a composed generic type.
fun test_singleton(): Array[Int] = singleton(12)

# Implicit generic instantiation with function argument
fun test_implicit_apply(): Int = apply(double, 12)

# Ambiguous generic instantiation
generic [T] fun fooz(x: T, y: T): Int
fun test_ambiguous_generic_instantiation(): Int = fooz(1, "lol")

# Instantiation is fine, wrong param
fun test_invalid_singleton(): Array[Int] = singleton("lol")

# Instantiation with wrong expected type 1
fun test_invalid_singleton_2(): Int = singleton("lol")

generic [T, U] struct Dummy {}

# Instantiation with wrong expected type 1
fun test_invalid_singleton_3(): Dummy[Int, String] = singleton("lol")

# Test rel operator
fun test_rel(a : Int): Bool = (a = 15) or (a = 18)

# Test that lambda params don't leak in their outside env
fun test_lambda_leak_support(a: (Int) -> Int, b: (Int) -> Int): Int
fun test_lambda_leak(): Int =
    test_lambda_leak_support(
        (xxx) => xxx * 2,
        (yyy) => xxx * yyy
#                ^ should be undefined
    )

# Implicit instantiations 2

# Test the built-in map function
fun test_map(ints: Array[Int]): Array[String] = ints.map((i) => convert_int[String](i))

# Test the built-in filtermap function
fun test_filtermap(ints: Array[Int]): Array[String] =
    ints.filtermap(
        (i) => convert_int[String](i),
        (i) => i = 0
    )

# Check that errors are handled correctly even for a call with a dotexpr prefix
fun test_handle_ce_error(arg : A): Int = arg.method_doesnt_exist()
# The method call should emit an error but not crash

# Check that conflicting actuals are correctly flagged
generic [T] fun mkPair(x: T): Pair[T, T]
fun test_conflicting_actuals(): Pair[Int, Bool] = mkPair(1)

#
# Basic subtyping
#
class Baseclass {
    fun foo(): Int
}

class Subclass : Baseclass {}

class Subsubclass : Subclass {}

class Subclass2 : Baseclass {}

fun test_subtyping_1 (a : Subclass): Baseclass = a
fun test_subtyping_2 (a : Subclass, b: Baseclass): Bool = (a = b)
fun test_subtyping_3 (a : Baseclass, b: Subsubclass): Bool =
  test_subtyping_2(b, b)
fun test_subtyping_invalid_1 (a : Subclass): Subclass2 = a
