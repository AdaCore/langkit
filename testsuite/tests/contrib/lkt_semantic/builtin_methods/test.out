Resolving test.lkt
==================
Id   <RefId "Node" test.lkt:1:14-1:18>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "TokenNode" test.lkt:1:30-1:39>
     references <TraitDecl "TokenNode" __prelude:45:10-47:2>

Id   <RefId "Node" test.lkt:3:25-3:29>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:3:32-3:36>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:3:32-3:36>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "parent" test.lkt:3:37-3:43>
     references <FunDecl "parent" __prelude:39:24-39:42>

Expr <RefId "parent" test.lkt:3:37-3:43>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Expr <DotExpr test.lkt:3:32-3:43>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "LexicalEnv" test.lkt:5:24-5:34>
     references <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "node" test.lkt:5:37-5:41>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:5:37-5:41>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "parent" test.lkt:5:42-5:48>
     references <FunDecl "parent" __prelude:39:24-39:42>

Expr <RefId "parent" test.lkt:5:42-5:48>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Expr <DotExpr test.lkt:5:37-5:48>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node_env" test.lkt:5:49-5:57>
     references <FunDecl "node_env" __prelude:40:14-40:40>

Expr <RefId "node_env" test.lkt:5:49-5:57>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <DotExpr test.lkt:5:37-5:57>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <CallExpr test.lkt:5:37-5:59>
     has type <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "LexicalEnv" test.lkt:6:24-6:34>
     references <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "node" test.lkt:6:37-6:41>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:6:37-6:41>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "parent_node" test.lkt:6:42-6:53>
     references <FunDecl "parent_node" test.lkt:3:5-3:43>

Expr <RefId "parent_node" test.lkt:6:42-6:53>
     has type <FunctionType "() -> Node" __prelude>

Expr <DotExpr test.lkt:6:37-6:53>
     has type <FunctionType "() -> Node" __prelude>

Expr <CallExpr test.lkt:6:37-6:55>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node_env" test.lkt:6:56-6:64>
     references <FunDecl "node_env" __prelude:40:14-40:40>

Expr <RefId "node_env" test.lkt:6:56-6:64>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <DotExpr test.lkt:6:37-6:64>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <CallExpr test.lkt:6:37-6:66>
     has type <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "Node" test.lkt:8:29-8:33>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:8:36-8:40>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:8:36-8:40>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "parent" test.lkt:8:41-8:47>
     references <FunDecl "parent" __prelude:39:24-39:42>

Expr <RefId "parent" test.lkt:8:41-8:47>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Expr <DotExpr test.lkt:8:36-8:47>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node_env" test.lkt:8:48-8:56>
     references <FunDecl "node_env" __prelude:40:14-40:40>

Expr <RefId "node_env" test.lkt:8:48-8:56>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <DotExpr test.lkt:8:36-8:56>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <CallExpr test.lkt:8:36-8:58>
     has type <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "get" test.lkt:8:59-8:62>
     references <FunDecl "get" __prelude:31:14-31:51>

Expr <RefId "get" test.lkt:8:59-8:62>
     has type <FunctionType "(Symbol) -> Array[Node]" __prelude>

Expr <DotExpr test.lkt:8:36-8:62>
     has type <FunctionType "(Symbol) -> Array[Node]" __prelude>

Id   <RefId "node" test.lkt:8:63-8:67>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:8:63-8:67>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "symbol" test.lkt:8:68-8:74>
     references <FunDecl "symbol" __prelude:46:24-46:44>

Expr <RefId "symbol" test.lkt:8:68-8:74>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Expr <DotExpr test.lkt:8:63-8:74>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Expr <CallExpr test.lkt:8:36-8:75>
     has type <InstantiatedGenericType "Array[Node]" __prelude:18:10-19:2>

Expr <NumLit test.lkt:8:77-8:78>
     has type <StructDecl "Int" __prelude:1:10-1:23>

Expr <NullCondCallExpr test.lkt:8:36-8:79>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "Node" test.lkt:9:29-9:33>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:9:36-9:40>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:9:36-9:40>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "parent_env" test.lkt:9:41-9:51>
     references <FunDecl "parent_env" test.lkt:6:5-6:66>

Expr <RefId "parent_env" test.lkt:9:41-9:51>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <DotExpr test.lkt:9:36-9:51>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <CallExpr test.lkt:9:36-9:53>
     has type <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "get" test.lkt:9:54-9:57>
     references <FunDecl "get" __prelude:31:14-31:51>

Expr <RefId "get" test.lkt:9:54-9:57>
     has type <FunctionType "(Symbol) -> Array[Node]" __prelude>

Expr <DotExpr test.lkt:9:36-9:57>
     has type <FunctionType "(Symbol) -> Array[Node]" __prelude>

Id   <RefId "node" test.lkt:9:58-9:62>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:9:58-9:62>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "symbol" test.lkt:9:63-9:69>
     references <FunDecl "symbol" __prelude:46:24-46:44>

Expr <RefId "symbol" test.lkt:9:63-9:69>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Expr <DotExpr test.lkt:9:58-9:69>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Expr <CallExpr test.lkt:9:36-9:70>
     has type <InstantiatedGenericType "Array[Node]" __prelude:18:10-19:2>

Expr <NumLit test.lkt:9:72-9:73>
     has type <StructDecl "Int" __prelude:1:10-1:23>

Expr <NullCondCallExpr test.lkt:9:36-9:74>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "Node" test.lkt:11:32-11:36>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:11:39-11:43>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:11:39-11:43>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "unit" test.lkt:11:44-11:48>
     references <FunDecl "unit" __prelude:42:14-42:38>

Expr <RefId "unit" test.lkt:11:44-11:48>
     has type <FunctionType "() -> AnalysisUnit" __prelude>

Expr <DotExpr test.lkt:11:39-11:48>
     has type <FunctionType "() -> AnalysisUnit" __prelude>

Expr <CallExpr test.lkt:11:39-11:50>
     has type <ClassDecl "AnalysisUnit" __prelude:35:10-37:2>

Id   <RefId "root" test.lkt:11:51-11:55>
     references <FunDecl "root" __prelude:36:24-36:40>

Expr <RefId "root" test.lkt:11:51-11:55>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Expr <DotExpr test.lkt:11:39-11:55>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "Node" test.lkt:13:24-13:28>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:13:31-13:35>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:13:31-13:35>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "node_env" test.lkt:13:36-13:44>
     references <FunDecl "node_env" __prelude:40:14-40:40>

Expr <RefId "node_env" test.lkt:13:36-13:44>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <DotExpr test.lkt:13:31-13:44>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <CallExpr test.lkt:13:31-13:46>
     has type <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "get_first" test.lkt:13:47-13:56>
     references <FunDecl "get_first" __prelude:32:14-32:50>

Expr <RefId "get_first" test.lkt:13:47-13:56>
     has type <FunctionType "(Symbol) -> Node" __prelude>

Expr <DotExpr test.lkt:13:31-13:56>
     has type <FunctionType "(Symbol) -> Node" __prelude>

Id   <RefId "node" test.lkt:13:57-13:61>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:13:57-13:61>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "symbol" test.lkt:13:62-13:68>
     references <FunDecl "symbol" __prelude:46:24-46:44>

Expr <RefId "symbol" test.lkt:13:62-13:68>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Expr <DotExpr test.lkt:13:57-13:68>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Expr <CallExpr test.lkt:13:31-13:69>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "Node" test.lkt:15:22-15:26>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:15:29-15:33>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:15:29-15:33>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "node_env" test.lkt:15:34-15:42>
     references <FunDecl "node_env" __prelude:40:14-40:40>

Expr <RefId "node_env" test.lkt:15:34-15:42>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <DotExpr test.lkt:15:29-15:42>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <CallExpr test.lkt:15:29-15:44>
     has type <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "env_node" test.lkt:15:45-15:53>
     references <FunDecl "env_node" __prelude:33:14-33:34>

Expr <RefId "env_node" test.lkt:15:45-15:53>
     has type <FunctionType "() -> Node" __prelude>

Expr <DotExpr test.lkt:15:29-15:53>
     has type <FunctionType "() -> Node" __prelude>

Expr <CallExpr test.lkt:15:29-15:55>
     has type <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "LexicalEnv" test.lkt:17:26-17:36>
     references <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "node" test.lkt:17:39-17:43>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:17:39-17:43>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "children_env" test.lkt:17:44-17:56>
     references <FunDecl "children_env" test.lkt:17:5-17:58>

Expr <RefId "children_env" test.lkt:17:44-17:56>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <DotExpr test.lkt:17:39-17:56>
     has type <FunctionType "() -> LexicalEnv" __prelude>

Expr <CallExpr test.lkt:17:39-17:58>
     has type <ClassDecl "LexicalEnv" __prelude:30:10-34:2>

Id   <RefId "String" test.lkt:19:24-19:30>
     references <StructDecl "String" __prelude:23:10-25:2>

Id   <RefId "Symbol" test.lkt:19:33-19:39>
     references <StructDecl "Symbol" __prelude:3:10-3:26>

Id   <RefId "s" test.lkt:19:42-19:43>
     references <FunArgDecl "s" test.lkt:19:20-19:30>

Expr <RefId "s" test.lkt:19:42-19:43>
     has type <StructDecl "String" __prelude:23:10-25:2>

Id   <RefId "to_symbol" test.lkt:19:44-19:53>
     references <FunDecl "to_symbol" __prelude:24:24-24:47>

Expr <RefId "to_symbol" test.lkt:19:44-19:53>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Expr <DotExpr test.lkt:19:42-19:53>
     has type <StructDecl "Symbol" __prelude:3:10-3:26>

Id   <RefId "Array" test.lkt:21:26-21:31>
     references <GenericDecl "Array" __prelude:18:10-19:2>

Id   <RefId "Node" test.lkt:21:32-21:36>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:21:40-21:44>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:21:40-21:44>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "parents" test.lkt:21:45-21:52>
     references <FunDecl "parents" __prelude:43:14-43:63>

Expr <RefId "parents" test.lkt:21:45-21:52>
     has type <FunctionType "(Bool) -> Array[Node]" __prelude>

Expr <DotExpr test.lkt:21:40-21:52>
     has type <FunctionType "(Bool) -> Array[Node]" __prelude>

Expr <CallExpr test.lkt:21:40-21:54>
     has type <InstantiatedGenericType "Array[Node]" __prelude:18:10-19:2>

Id   <RefId "Array" test.lkt:22:26-22:31>
     references <GenericDecl "Array" __prelude:18:10-19:2>

Id   <RefId "Node" test.lkt:22:32-22:36>
     references <ClassDecl "Node" __prelude:38:10-44:2>

Id   <RefId "node" test.lkt:22:40-22:44>
     references <NodeDecl "node" test.lkt:1:1-24:2>

Expr <RefId "node" test.lkt:22:40-22:44>
     has type <ClassDecl "Name" test.lkt:1:1-24:2>

Id   <RefId "parents" test.lkt:22:45-22:52>
     references <FunDecl "parents" __prelude:43:14-43:63>

Expr <RefId "parents" test.lkt:22:45-22:52>
     has type <FunctionType "(Bool) -> Array[Node]" __prelude>

Expr <DotExpr test.lkt:22:40-22:52>
     has type <FunctionType "(Bool) -> Array[Node]" __prelude>

Id   <RefId "with_self" test.lkt:22:53-22:62>
     references <FunArgDecl "with_self" __prelude:43:26-43:49>

Expr <RefId "with_self" test.lkt:22:53-22:62>
     has type <EnumTypeDecl "Bool" __prelude:5:16-7:2>

Id   <RefId "false" test.lkt:22:63-22:68>
     references <EnumLitDecl "false" __prelude:6:10-6:15>

Expr <RefId "false" test.lkt:22:63-22:68>
     has type <EnumTypeDecl "Bool" __prelude:5:16-7:2>

Expr <CallExpr test.lkt:22:40-22:69>
     has type <InstantiatedGenericType "Array[Node]" __prelude:18:10-19:2>

